src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,387 @@
+
+                Expression compare = null;
+                ODataEnumValue enumValue = value as ODataEnumValue;
+                if (enumValue != null)
+                {
+                    value = enumValue.Value;
+                }
+                Expression constant = parameterizeConstant ? LinqParameterContainer.Parameterize(value.GetType(), value) : Expression.Constant(value);
+                if (directionMap.ContainsKey(key))
+                {
+                    compare = directionMap[key] == OrderByDirection.Descending ? binder.CreateBinaryExpression(BinaryOperatorKind.LessThan, property, constant, true) : binder.CreateBinaryExpression(BinaryOperatorKind.GreaterThan, property, constant, true);
+                }
+                else
+                {
+                    compare = binder.CreateBinaryExpression(BinaryOperatorKind.GreaterThan, property, constant, true);
+                }
+
+                if (firstProperty)
+                {
+                    lastEquality = binder.CreateBinaryExpression(BinaryOperatorKind.Equal, property, constant, true);
+                    where = compare;
+                    firstProperty = false;
+                }
+                else
+                {
+                    Expression condition = Expression.AndAlso(lastEquality, compare);
+                    where = where == null ? condition : Expression.OrElse(where, condition);
+                    lastEquality = Expression.AndAlso(lastEquality, binder.CreateBinaryExpression(BinaryOperatorKind.Equal, property, constant, true));
+                }
+            }
+
+            Expression whereLambda = Expression.Lambda(where, param);
+            return ExpressionHelpers.Where(query, whereLambda, query.ElementType);
+        }
+
+        /// <summary>
+        /// Generates a dictionary with property name and property values specified in the skiptoken value.
+        /// </summary>
+        /// <param name="value">The skiptoken string value.</param>
+        /// <param name="context">The <see cref="ODataQueryContext"/> which contains the <see cref="IEdmModel"/> and some type information</param>
+        /// <returns>Dictionary with property name and property value in the skiptoken value.</returns>
+        protected static IDictionary<string, object> PopulatePropertyValuePairs(string value, ODataQueryContext context)
+        {
+            Contract.Assert(context != null);
+
+            IDictionary<string, object> propertyValuePairs = new Dictionary<string, object>();
+            string[] keyValues = value.Split(CommaDelimiter);
+            foreach (string keyAndValue in keyValues)
+            {
+                string[] pieces = keyAndValue.Split(new char[] { propertyDelimiter }, 2);
+                if (pieces.Length > 1 && !String.IsNullOrWhiteSpace(pieces[1]))
+                {
+                    object propValue = null;
+                    if (pieces[1].StartsWith("'enumType'"))
```

Comment:
>if (pieces[1].StartsWith("'enumType'")) [](start = 20, length = 39)

why do we have this special test? 

I think it should use the EdmTypeKind to identify whether it's enum or other primitive type?

Reply:
use the type in line 288 to test?

---
In reply to: [282591059](https://github.com/OData/WebApi/pull/1724#discussion_r282591059) [](ancestors = 282591059)

Reply:
GetTypeReferenceOfProperty is an expensive operation. We don't want to do it for all the types. It is only required for enum.

---
In reply to: [282591584](https://github.com/OData/WebApi/pull/1724#discussion_r282591584) [](ancestors = 282591584,282591059)